clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -5413,7 +5413,7 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
            Triple.getSubArch() == llvm::Triple::SPIRSubArch_gen) ||
           Triple.isNVPTX() || Triple.isAMDGCN()) {
         StringRef Device = JA.getOffloadingArch();
-        if (!Device.empty()) {
+        if (!Device.empty() && !SYCL::gen::getGenDeviceMacro(Device).empty()) {
```

Comment:
Should we add a test for this circumstance?  Also, would it be reasonable to emit some kind of diagnostic that the macro isn't being generated due to an issue with the device value?

Reply:
> Should we add a test for this circumstance?

Could you advise on the best place to put a test? `clang/test/Driver/sycl-oneapi-gpu.cpp`?

> Also, would it be reasonable to emit some kind of diagnostic that the macro isn't being generated due to an issue with the device value?

I don't think macro alone warrants this. It's prone to false positives (what if the code don't use these macros at all?), and occurs at the wrong place (it's not the problem of the user that they have gfx1036, it's the problem of the developer who wrote code relying a macro `__SYCL_TARGET_AMD_GPU_GFX1036__`). Warning about unsupported macros in the code (`__SYCL_TARGET_*` will never be set) independently of the target architecture could will be better: it will be emitted if and only if there is an actual bug in the code, but this is already covered by the undefined macro warning (since, if the macro is supported, it will be defined to 0, https://github.com/intel/llvm/blob/de92299c2c09d626dcbc633c83e259d828220d03/sycl/doc/design/DeviceIf.md#L196).

We could, more generally, warn about "unsupported" devices, the way Clang warns about "unsupported" CUDA versions (`-Wunknown-cuda-version`). I don't think it fits the scope of this PR, though.

Reply:
I agree with @al42and . If you read the [DeviceIf.md](https://github.com/intel/llvm/blob/de92299c2c09d626dcbc633c83e259d828220d03/sycl/doc/design/DeviceIf.md), it is quite clear IMO:

"These macros are an internal implementation detail, so they should not be documented to users, and user code should not make use of them."

@al42and I think the answer to your question is that the appropriate place to place such a test is in clang/test/Driver: I guess you just want something similar to this: https://github.com/JackAKirk/llvm/blob/2b4b45af5a1ae0d23bcb632cc4588faecedc1956/clang/test/Driver/sycl-fno-libspirv-warn.cpp

except you want to check there are no warnings or errors when compiling for a device that e.g doesn't have a macro assigned but is supported by the driver.

Reply:
> except you want to check there are no warnings or errors when compiling for a device that e.g doesn't have a macro assigned but is supported by the driver.

That's going to be an unreliable test and in need of regular updates. Hopefully, such architectures will eventually get assigned a flag :)

I think a better test is to check that when compiling for `gfx0000` or `sm_00` we get a reasonable error instead of an ICE.

Reply:
> I think a better test is to check that when compiling for gfx0000 or sm_00 we get a reasonable error instead of an ICE.

Unfortunately, the validity of the architecture is checked before the broken compiler pass is invoked, so this idea will not work. 

Instead, I'm using an old architecture (e.g., `sm_30`), that is valid but is not and will not be supported.

Reply:
Looks like NVIDIA devive libraries are not as reliable as I thought they are. Changed the test to use an old AMD architecture (`gfx600`, supported by LLVM, not listed in `clang/lib/Driver/ToolChains/SYCL.cpp`) instead.

Reply:
These tests are harder than I expected :sweat_smile: Hopefully I'll get "Build + LIT" to pass this time with 081d821f0d42799b9fefdc4dc2bbdf52903d5d28.

The timeout of `Basic/vector/int-convert.cpp` in [e2e / Intel GEN12 Graphics with Level Zero](https://github.com/intel/llvm/actions/runs/7017785962/job/19094128283#logs) should be unrelated, though.